### ISSUE TITLE
Bug: Missing Collection Metadata Account in BurnNft Ctx Struct for Metaplex Token Metadata burn nft instruction

### ISSUE THREAD

**Issue by nabeel99** [coral-xyz/anchor#2804]
#### Context:
While working with the Anchor framework spl wrapper over metaplex token metadata, I've identified a potential issue in the `BurnNft` struct, specifically related to the handling of NFT collections.

#### Problem:
The current `BurnNft` struct definition lacks a direct reference to the `collection_metadata` account. This omission complicates the process of burning NFTs that are part of a collection and have associated collection metadata. As it stands, users are required to retrieve the `collection_metadata` account indirectly from extra accounts and pass it to the Metaplex token metadata burn NFT instruction. This approach is counterintuitive and may not align with the intended design of the wrapper.

#### Current Struct Definition:
```rust
#[derive(Accounts)]
pub struct BurnNft<'info> {
    pub metadata: AccountInfo<'info>,
    pub owner: AccountInfo<'info>,
    pub mint: AccountInfo<'info>,
    pub token: AccountInfo<'info>,
    pub edition: AccountInfo<'info>,
    pub spl_token: AccountInfo<'info>,
}
```

#### Suggested Change:
To bring the struct definition as to what  Metaplex's token metadata burnNFT instruction expect, especially for NFTs that are part of a collection, I propose adding an additional field to the `BurnNft` struct:

```rust
pub collection_metadata: AccountInfo<'info>,
```

#### Impact:
Without this field, the burn NFT instruction fails for NFTs that are part of a collection and have a corresponding collection metadata NFT. The current method only allows passing explicitly mentioned accounts in the `ctx` to `ToAccountInfos::to_account_infos(&ctx)`, which does not include `collection_metadata` by default.

#### Proposed Solution:
1. **Add `collection_metadata` to the `BurnNft` struct**: This will enable users to pass the collection metadata account directly from `ctx.accounts`, aligning with intuitive usage patterns.
2. **Update Documentation**: If the exclusion of `collection_metadata` was intentional, it's crucial to document this design choice and provide guidance on how users should handle collection metadata accounts under the current structure.

#### Code Snippet for Reference:
```
let ix = mpl_token_metadata::instructions::BurnNft {
    collection_metadata, # public key of collection metadata passed
    master_edition_account: *ctx.accounts.edition.key,
    metadata: *ctx.accounts.metadata.key,
    mint: *ctx.accounts.mint.key,
    owner: *ctx.accounts.owner.key,
    spl_token_program: *ctx.accounts.spl_token.key,
    token_account: *ctx.accounts.token.key,
}.instruction();
solana_program::program::invoke_signed(
    &ix,
    &ToAccountInfos::to_account_infos(&ctx), # collection metadata not declared as a field hence the account itself wdnt be passed
    ctx.signer_seeds,
).map_err(Into::into)
```

**Comment by nabeel99** [comment]
if my finding is correct and this was unintentional, i would be more than happy to contribute the fix for this, i just needed validation if my understanding of this is correct.

**Comment by acheroncrypto** [comment]
> if my finding is correct and this was unintentional, i would be more than happy to contribute the fix for this, i just needed validation if my understanding of this is correct.

This looks to be intentional since the account is not required to be passed in, @febo can correct me if I'm wrong on that. You can pass optional CPI accounts with `ctx.with_remaining_accounts` when necessary.

> #### Proposed Solution:
> 1. **Add `collection_metadata` to the `BurnNft` struct**: This will enable users to pass the collection metadata account directly from `ctx.accounts`, aligning with intuitive usage patterns.
> 2. **Update Documentation**: If the exclusion of `collection_metadata` was intentional, it's crucial to document this design choice and provide guidance on how users should handle collection metadata accounts under the current structure.

Documenting the usage would be great!

**Comment by febo** [comment]
Yes, the `collection_metadata` is an optional account. The (current) `BurnNft<'info>` was defined before we had optional accounts on Anchor. So the reason that it is not on the struct is because there was no way to specify that it was an optional account at that time.

There is a trade-off of adding the (optional) collection metadata to the struct. If it is added, it will require the use of `Option<AccountInfo<'info>>`, which I am not sure whether it is desirable to "force" the parent instruction to use optional accounts or not. As an alternative, it is possible to add a `collection_metadata: Option<AccountInfo<'info>>` to the parent instruction to avoid having to "fetch" the collection metadata from the remaining accounts – e.g.:
```rust
#[derive(Accounts)]
pub struct BurnMyNft<'info> {
   ...
   burn_accounts: BurnNft<'info>,
   collection_metadata:  Option<AccountInfo<'info>>,
}
```
Then you can pass `ctx.accounts.collection_metadata` directly to the burn wrapper.

**Comment by nabeel99** [comment]
> > if my finding is correct and this was unintentional, i would be more than happy to contribute the fix for this, i just needed validation if my understanding of this is correct.
> 
> 
> 
> This looks to be intentional since the account is not required to be passed in, @febo can correct me if I'm wrong on that. You can pass optional CPI accounts with `ctx.with_remaining_accounts` when necessary.
> 
> 
> 
> > #### Proposed Solution:
> 
> > 1. **Add `collection_metadata` to the `BurnNft` struct**: This will enable users to pass the collection metadata account directly from `ctx.accounts`, aligning with intuitive usage patterns.
> 
> > 2. **Update Documentation**: If the exclusion of `collection_metadata` was intentional, it's crucial to document this design choice and provide guidance on how users should handle collection metadata accounts under the current structure.
> 
> 
> 
> Documenting the usage would be great!

Am aware this is one possible solutions but its kinda not exactly convenient to use it this way , i was thinking along the lines of what @febo  suggested too, so that it does not sacrifice the ergonomic/dev ux , happy to either go with the documentation approach or what febo suggested and make the relevant pr going forward.

**Comment by acheroncrypto** [comment]
> Am aware this is one possible solutions but its kinda not exactly convenient to use it this way , i was thinking along the lines of what @febo suggested too, so that it does not sacrifice the ergonomic/dev ux , happy to either go with the documentation approach or what febo suggested and make the relevant pr going forward.

You can use optional accounts in your instruction to avoid having to get the collection metadata account from the remaining accounts like @febo suggested.

However, it would be better to go with the documentation approach for `anchor-spl` since optional accounts are not available by default.

**Comment by Shiva953** [comment]
Is this implemented now? If not, I would be happy to take this up

**Comment by nabeel99** [comment]
> Is this implemented now? If not, I would be happy to take this up

Hey i am on it with the decided fix.

**Comment by nabeel99** [comment]
> Yes, the `collection_metadata` is an optional account. The (current) `BurnNft<'info>` was defined before we had optional accounts on Anchor. So the reason that it is not on the struct is because there was no way to specify that it was an optional account at that time.
> 
> There is a trade-off of adding the (optional) collection metadata to the struct. If it is added, it will require the use of `Option<AccountInfo<'info>>`, which I am not sure whether it is desirable to "force" the parent instruction to use optional accounts or not. As an alternative, it is possible to add a `collection_metadata: Option<AccountInfo<'info>>` to the parent instruction to avoid having to "fetch" the collection metadata from the remaining accounts – e.g.:
> 
> ```rust
> #[derive(Accounts)]
> pub struct BurnMyNft<'info> {
>    ...
>    burn_accounts: BurnNft<'info>,
>    collection_metadata:  Option<AccountInfo<'info>>,
> }
> ```
> 
> Then you can pass `ctx.accounts.collection_metadata` directly to the burn wrapper.
Needed a bit of clarification @acheroncrypto ,
I am a bit confused as to how @febo  solution works, since if i want to use the wrapper, i have to construct a `BurnNft` CpiContext type , if the `BurnNft` struct does not take the optional `collection_metadata` how would one call this wrapper function with an nft that has a `collection_metadata` account, is the proposed solution to document the fact this wrapper wont work with a nft if it has a collection_metadata account  but then this begs the question what is the point of having `collection_metadata : Option<Pubkey>` as a param to the `burn_nft` function ?

**Comment by nabeel99** [comment]
> > Yes, the `collection_metadata` is an optional account. The (current) `BurnNft<'info>` was defined before we had optional accounts on Anchor. So the reason that it is not on the struct is because there was no way to specify that it was an optional account at that time.
> > There is a trade-off of adding the (optional) collection metadata to the struct. If it is added, it will require the use of `Option<AccountInfo<'info>>`, which I am not sure whether it is desirable to "force" the parent instruction to use optional accounts or not. As an alternative, it is possible to add a `collection_metadata: Option<AccountInfo<'info>>` to the parent instruction to avoid having to "fetch" the collection metadata from the remaining accounts – e.g.:
> > ```rust
> > #[derive(Accounts)]
> > pub struct BurnMyNft<'info> {
> >    ...
> >    burn_accounts: BurnNft<'info>,
> >    collection_metadata:  Option<AccountInfo<'info>>,
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then you can pass `ctx.accounts.collection_metadata` directly to the burn wrapper.
> 
> Needed a bit of clarification @acheroncrypto , I am a bit confused as to how @febo solution works, since if i want to use the wrapper, i have to construct a `BurnNft` CpiContext type , if the `BurnNft` struct does not take the optional `collection_metadata` how would one call this wrapper function with an nft that has a `collection_metadata` account, is the proposed solution to document the fact this wrapper wont work with a nft if it has a collection_metadata account but then this begs the question what is the point of having `collection_metadata : Option<Pubkey>` as a param to the `burn_nft` function ? The function in question here is the following :
> 
> ```
> pub fn burn_nft<'info>(
>     ctx: CpiContext<'_, '_, '_, 'info, BurnNft<'info>>,
>     collection_metadata: Option<Pubkey>,
> ) -> Result<()> {
>     let ix = mpl_token_metadata::instructions::BurnNft {
>         collection_metadata,
>         master_edition_account: *ctx.accounts.edition.key,
>         metadata: *ctx.accounts.metadata.key,
>         mint: *ctx.accounts.mint.key,
>         owner: *ctx.accounts.owner.key,
>         spl_token_program: *ctx.accounts.spl_token.key,
>         token_account: *ctx.accounts.token.key,
>     }
>     .instruction();
>     solana_program::program::invoke_signed(
>         &ix,
>         &ToAccountInfos::to_account_infos(&ctx),
>         ctx.signer_seeds,
>     )
>     .map_err(Into::into)
> }
> ```

nvm i read the source code for `CpiContext` and understood how `CpiContext with_remaining_accounts` is implemented and how `ToAccountInfos` is impl for `CpiContext` works, it now makes sense to me why y'all suggested taking the collection_metadata account in the parent instruction and add it to the `BurnNft` CpiContext Struct via the `with_remaining_accounts` method on`CpiContext`
So to recap the solution for this sort of scenario is as follows :

- Add the collection metadata account in the parent instruction
- create the burnnft cpicontext
- call the created cpicontext with remaining accounts method and pass in the collection_metadata as input to it
- then pass the cpicontext of burn nft to the wrapper function as normal.

@febo @acheroncrypto  is the above solution to the issue , sufficient to add to the doc ?
will open a pr with this scenario documented.

**Comment by acheroncrypto** [comment]
So to recap the solution for this sort of scenario is as follows :
> 
> * Add the collection metadata account in the parent instruction
> * create the burnnft cpicontext
> * call the created cpicontext with remaining accounts method and pass in the collection_metadata as input to it
> * then pass the cpicontext of burn nft to the wrapper function as normal.
> 
> @febo @acheroncrypto is the above solution to the issue sufficient and i should proceed with opening up the pr documenting this scenario ?

Yes, sounds good to me.

**Comment by nabeel99** [comment]
closed via : #2819
